M2/Macaulay2/d/regex.dd

```diff
@@ -0,0 +1,242 @@
+	else WrongArgString(2)
+	else WrongArgString(1))
+    else if length(s) == 6 then (
+	when s.0 is regexp:stringCell do
+	when s.1 is start:ZZcell do if !isInt(start) then WrongArgSmallInteger(2) else
+	when s.2 is range:ZZcell do if !isInt(range) then WrongArgSmallInteger(3) else
+	when s.3 is text:stringCell do
+	when s.4 is regexFlags:ZZcell do if !isInt(regexFlags) then WrongArgSmallInteger(5) else
+	when s.5 is matchFlags:ZZcell do if !isInt(matchFlags) then WrongArgSmallInteger(6) else (
+	    istart := toInt(start);
+	    irange := toInt(range);
+	    Ccode(returns, "try {");
+	    r := search(regexp.v, istart, irange, text.v, toInt(regexFlags), toInt(matchFlags));
+	    Ccode(returns, "
+		} catch (std::exception& e) {
+		    fprintf(stderr, \"-- %s\\n\", e.what());
+		    return ", buildErrorPacket("unable to compile regular expression"), ";
+		}");
+	    if length(r) != 0 then toPairs(r)
+	    else nullE)
+	else WrongArgZZ(6)
+	else WrongArgZZ(5)
+	else WrongArgString(4)
+	else WrongArgZZ(3)
+	else WrongArgZZ(2)
+	else WrongArgString(1))
+    else WrongNumArgs(2,6)
+    else WrongNumArgs(2,6));
+setupfun("regex", regexSearch).Protected = false; -- will be overloaded in m2/regex.m2
+
+regexReplace(e:Expr):Expr := (
+    when e is s:Sequence do
+    if length(s) == 3 then (
+	when s.0 is regexp:stringCell do
+	when s.1 is replacement:stringCell do
+	when s.2 is text:stringCell do (
+	    Ccode(returns, "try {");
+	    r := replace(regexp.v, 0, length(text.v), replacement.v, text.v, defaultRegexFlags, defaultMatchFlags);
+	    Ccode(returns, "
+		} catch (std::exception& e) {
+		    fprintf(stderr, \"-- %s\\n\", e.what());
+		    return ", buildErrorPacket("unable to compile regular expression"), ";
+		}");
+	    toExpr(r))
+	else WrongArgString(3)
+	else WrongArgString(2)
+	else WrongArgString(1))
+    else if length(s) == 5 then (
+	when s.0 is regexp:stringCell do
+	when s.1 is replacement:stringCell do
+	when s.2 is text:stringCell do
+	when s.3 is regexFlags:ZZcell do if !isInt(regexFlags) then WrongArgSmallInteger(4) else
+	when s.4 is matchFlags:ZZcell do if !isInt(matchFlags) then WrongArgSmallInteger(5) else (
+	    Ccode(returns, "try {");
+	    r := replace(regexp.v, 0, length(text.v), replacement.v, text.v, toInt(regexFlags), toInt(matchFlags));
+	    Ccode(returns, "
+		} catch (std::exception& e) {
+		    fprintf(stderr, \"-- %s\\n\", e.what());
+		    return ", buildErrorPacket("unable to compile regular expression"), ";
+		}");
+	    toExpr(r))
+	else WrongArgZZ(5)
+	else WrongArgZZ(4)
+	else WrongArgString(3)
+	else WrongArgString(2)
+	else WrongArgString(1))
+    else WrongNumArgs(3,5)
+    else WrongNumArgs(3,5));
+setupfun("replace", regexReplace).Protected = false; -- will be overloaded in m2/regex.m2
+
+regexSeparate(e:Expr):Expr := (
+    when e is text:stringCell do (
+	Ccode(returns, "try {");
+	r := separate("\r?\n", 0, length(text.v), text.v, defaultRegexFlags, defaultMatchFlags);
+	Ccode(returns, "
```

Comment:
The file c/README clarifies the proper use of "returns", which means that flow of control never passes to the next statement, so all of these in Ccode should be changed to "void".

Reply:
Is there a difference in generated code? Eitherway, done. I also turned all those `Regex$` variables into a hashtable.